File: bibgraph/library/graph.py
"""Build the citation graph + API projections from enriched library works.

Nodes = papers (saved works + suggested neighbours), edges = citation links.
Edges come from two sources, unioned:

* **OpenAlex** ``referenced_works`` — the broad network (saved→neighbour,
  neighbour→neighbour, and saved→saved when OpenAlex matched the bibliography);
* **offline** — each ingested paper's own parsed bibliography matched against
  the saved set (guarantees the saved↔saved edges regardless of OpenAlex).

Suggested neighbours are the most-cited works referenced by the library, capped
so the payload stays small; the frontend's depth/count sliders filter further.
"""

from __future__ import annotations

import logging

from .seed import doc_reference_ids
from .sources.openalex import OpenAlex
from .store import LibraryStore, Work, display_authors, norm_arxiv, norm_doi, norm_title, slug

log = logging.getLogger("bibgraph.library.graph")

MAX_SUGGEST = 120


# --------------------------------------------------------------------------- #
# Enrichment
# --------------------------------------------------------------------------- #

def _suffix(n: int) -> str:
    """1→'a', 26→'z', 27→'aa', … (always [a-z], no overflow past 'z')."""
    s, k = "", n
    while k > 0:
        k, r = divmod(k - 1, 26)
        s = chr(ord("a") + r) + s
    return s


def _cite_key(w: Work, used: set[str]) -> str:
    fam = (w.authors[0] if w.authors else slug(w.title, 12)).lower()
    fam = "".join(ch for ch in fam if ch.isalnum()) or "ref"
    base = f"{fam}{w.year or ''}"
    key, n = base, 0
    while key in used:
        n += 1
        key = base + _suffix(n)
    used.add(key)
    return key


# Metadata enrichment now lives in the acquisition layer's resolution chain
# (ADS▸Crossref▸OpenAlex); see bibgraph.acquire.resolve / bibgraph.acquire.run.
# This module keeps _cite_key (shared) plus offline edge + graph assembly.


# --------------------------------------------------------------------------- #
# Offline edges (saved ↔ saved from the ingested bibliographies)
# --------------------------------------------------------------------------- #

def offline_edges(store: LibraryStore) -> set[tuple[str, str]]:
    key2wid: dict[str, str] = {}
    for w in store.works:
        for k in w.identity_keys():
            key2wid[k] = w.id
    edges: set[tuple[str, str]] = set()
    for w in store.works:
        for doc_id in w.doc_ids:
            for r in doc_reference_ids(doc_id):
                cks: set[str] = set()
                if r.get("doi"):
                    cks.add("doi:" + (norm_doi(r["doi"]) or ""))
                if r.get("arxiv_id"):
                    cks.add("arxiv:" + (norm_arxiv(r["arxiv_id"]) or ""))
                if r.get("title"):
                    nt = norm_title(r["title"])
                    if len(nt) > 12:  # avoid spurious short-title collisions
                        cks.add("title:" + nt)
                for ck in cks:
                    tgt = key2wid.get(ck)
                    if tgt and tgt != w.id:
                        edges.add((w.id, tgt))
    return edges


# --------------------------------------------------------------------------- #
# Graph assembly
# --------------------------------------------------------------------------- #

def _saved_node(w: Work) -> dict:
    return {
        "id": w.id,
        "ref": w.id,
        "y": w.year or 2024,
        "c": w.cited_by_count or 0,
        "a": display_authors(w.authors) or (w.title[:24] if w.title else w.id),
        "v": w.venue or "",
        "t": w.title or w.id,
        "doi": w.doi,
        "arxiv_id": w.arxiv_id,
        "doc_id": w.doc_ids[0] if w.doc_ids else None,
    }


def _suggested_node(m: dict) -> dict:
    return {
        "id": "oa:" + m["openalex_id"],
        "y": m["year"],
        "c": m["cited_by_count"] or 0,
        "a": display_authors(m["authors"]) or (m["title"][:24] if m["title"] else m["openalex_id"]),
        "v": m["venue"] or "",
        "t": m["title"] or m["openalex_id"],
        "doi": m["doi"],
    }


def build_graph(store: LibraryStore, oa: OpenAlex) -> dict:
    oaid2node: dict[str, str] = {}
    nodes: list[dict] = []
    for w in store.works:
        nodes.append(_saved_node(w))
        if w.openalex_id:
            oaid2node[w.openalex_id] = w.id

    saved_oa = {w.openalex_id for w in store.works if w.openalex_id}
    pool: set[str] = set()
    for w in store.works:
        pool.update(w.referenced_works or [])
    pool -= saved_oa

    meta = oa.fetch_many(sorted(pool)) if pool else {}
    cands = sorted(
        (m for m in meta.values() if m.get("year")),
        key=lambda m: m.get("cited_by_count") or 0,
        reverse=True,
    )[:MAX_SUGGEST]
    for c in cands:
        nid = "oa:" + c["openalex_id"]
        oaid2node[c["openalex_id"]] = nid
        nodes.append(_suggested_node(c))

    links: set[tuple[str, str]] = set()

    def add_refs(refs: list[str] | None, src_node: str) -> None:
        for r in refs or []:
            tgt = oaid2node.get(r)
            if tgt and tgt != src_node:
                links.add((src_node, tgt))

    for w in store.works:
        add_refs(w.referenced_works, w.id)
    for c in cands:
        add_refs(c.get("referenced_works"), "oa:" + c["openalex_id"])
    links |= offline_edges(store)

    # keep only nodes that are saved or actually connected
    connected = {a for a, _ in links} | {b for _, b in links}
    saved_ids = {w.id for w in store.works}
    kept = [n for n in nodes if n["id"] in saved_ids or n["id"] in connected]
    kept_ids = {n["id"] for n in kept}
    links = {(a, b) for a, b in links if a in kept_ids and b in kept_ids}

    log.info("graph: %d nodes (%d saved) · %d links", len(kept), len(saved_ids), len(links))
    return {"nodes": kept, "links": [[a, b] for a, b in sorted(links)]}


# --------------------------------------------------------------------------- #
# API projections
# --------------------------------------------------------------------------- #

def work_to_ref(w: Work) -> dict:
    # required fields (the TS LibraryRef declares these non-optional) are always
    # present — never stripped — so the frontend can rely on e.g. r.authors.
    ref = {
        "id": w.id,
        "title": w.title or w.id,
        "authors": display_authors(w.authors),
        "year": w.year or 0,
        "venue": w.venue or "",
        "type": w.type,
        "cite": w.cite_key or slug(w.id),
        "tags": w.tags,
        "pdf": bool(w.doc_ids),
        "read": w.read,
        "note": bool(w.note),
        "star": w.star,
    }
    acq = w.acquisition or {}
    optional = {
        "abstract": w.abstract,
        "doi": w.doi,
        "arxiv_id": w.arxiv_id,
        "doc_id": w.doc_ids[0] if w.doc_ids else None,
        "citedBy": w.cited_by_count,
        "label": w.label,
        "journal": w.journal,
        # acquisition: which full-text source is planned + whether it's ready now
        "source": acq.get("chosen"),
        "sourceLabel": acq.get("chosen_label"),
        "sourceStatus": acq.get("status"),
        "sourceReady": acq.get("ready"),
        # bot-walled & no auto source → the UI offers a PDF upload
        "needs_upload": acq.get("needs_upload") or None,
        # resolution: which provider supplied the citation count (ads/crossref/openalex)
        "resolvedBy": (w.resolution or {}).get("count"),
    }
    ref.update({k: v for k, v in optional.items() if v is not None and v != ""})
    return ref

File: packages/core/src/library/graph.ts
/**
 * Build the citation graph + API projections from enriched library works
 * (bibgraph/library/graph.py).
 *
 * **Stage 14: the Library graph is saved-only** (the global suggested-node
 * recommendation architecture is retired — ADS Discovery replaces it). Nodes
 * are exactly the saved works; edges are real citation links between them,
 * unioned from two local sources:
 * - **OpenAlex** `referenced_works` already stored on the works (no network);
 * - **offline** — each ingested paper's own parsed bibliography matched
 *   against the saved set.
 *
 * After enrichment the build is therefore pure/local: no candidate pool, no
 * metadata fetch for suggestions, no cited_by_count ranking, no cap.
 *
 * Port notes (bug-for-bug):
 * - Python's `_saved_node` emits explicit `null`s (the graph JSON is dumped
 *   un-compacted) — kept.
 * - Python's `sorted(links)` orders (a, b) tuples lexicographically; the TS
 *   port compares element-wise to the same effect.
 */

import {
  CURRENT_GRAPH_VERSION,
  type GraphData,
  type GraphNode,
  type LibraryRef,
} from "@argelanderspace/contracts";
import { pyOr } from "../documents/pyregex.js";
import { docReferenceIds } from "./seed.js";
import {
  displayAuthors,
  identityKeys,
  type LibraryStore,
  normArxiv,
  normDoi,
  normTitle,
  slug,
  type Work,
} from "./store.js";

// --------------------------------------------------------------------------- //
// Enrichment helpers
// --------------------------------------------------------------------------- //

/** 1→'a', 26→'z', 27→'aa', … (always [a-z], no overflow past 'z'). */
function suffix(n: number): string {
  let s = "";
  let k = n;
  while (k > 0) {
    const r = (k - 1) % 26;
    k = Math.floor((k - 1) / 26);
    s = String.fromCharCode(97 + r) + s;
  }
  return s;
}

/** `_cite_key`: family-name + year base, disambiguated with {@link suffix}. */
export function citeKey(w: Work, used: Set<string>): string {
  let fam = (w.authors.length > 0 ? (w.authors[0] as string) : slug(w.title, 12)).toLowerCase();
  // Python str.isalnum() is Unicode-aware
  fam = [...fam].filter((ch) => /[\p{L}\p{N}]/u.test(ch)).join("") || "ref";
  const base = `${fam}${w.year || ""}`;
  let key = base;
  let n = 0;
  while (used.has(key)) {
    n += 1;
    key = base + suffix(n);
  }
  used.add(key);
  return key;
}

/**
 * Stage 12 cite-key assignment: a work WITH an ADS bibcode takes the bibcode
 * itself as its key — the bibcode is the permanent ADS identifier, so the key
 * needs no family/year disambiguation and is stable by design. Works without a
 * bibcode keep the family+year scheme. An assigned key NEVER changes (Stage 7
 * contract), including works that only gain a bibcode on a later rebuild.
 */
export function assignCiteKey(w: Work, used: Set<string>): string {
  if (w.bibcode && !used.has(w.bibcode)) {
    used.add(w.bibcode);
    return w.bibcode;
  }
  return citeKey(w, used);
}

// --------------------------------------------------------------------------- //
// Offline edges (saved ↔ saved from the ingested bibliographies)
// --------------------------------------------------------------------------- //

/** Directed edge between two node ids. */
export type Edge = readonly [string, string];

function edgeKey([a, b]: Edge): string {
  return `${a}\u0001${b}`;
}

/** Compare edges as Python compares tuples: by first, then second element. */
export function compareEdges(x: Edge, y: Edge): number {
  if (x[0] !== y[0]) return x[0] < y[0] ? -1 : 1;
  if (x[1] !== y[1]) return x[1] < y[1] ? -1 : 1;
  return 0;
}

export function offlineEdges(store: LibraryStore, outputDir: string): Edge[] {
  const key2wid = new Map<string, string>();
  for (const w of store.works) {
    for (const k of identityKeys(w)) key2wid.set(k, w.id);
  }
  const seen = new Set<string>();
  const edges: Edge[] = [];
  for (const w of store.works) {
    for (const docId of w.doc_ids) {
      for (const r of docReferenceIds(docId, outputDir)) {
        const cks = new Set<string>();
        const rDoi = typeof r.doi === "string" ? r.doi : null;
        if (rDoi) cks.add(`doi:${normDoi(rDoi) || ""}`);
        const rArxiv = typeof r.arxiv_id === "string" ? r.arxiv_id : null;
        if (rArxiv) cks.add(`arxiv:${normArxiv(rArxiv) || ""}`);
        const rTitle = typeof r.title === "string" ? r.title : null;
        if (rTitle) {
          const nt = normTitle(rTitle);
          if (nt.length > 12) cks.add(`title:${nt}`); // avoid spurious short-title collisions
        }
        for (const ck of cks) {
          const tgt = key2wid.get(ck);
          if (tgt && tgt !== w.id) {
            const e: Edge = [w.id, tgt];
            if (!seen.has(edgeKey(e))) {
              seen.add(edgeKey(e));
              edges.push(e);
            }
          }
        }
      }
    }
  }
  return edges;
}

// --------------------------------------------------------------------------- //
// Graph assembly
// --------------------------------------------------------------------------- //

function savedNode(w: Work): GraphNode {
  return {
    id: w.id,
    ref: w.id,
    y: w.year || 2024,
    c: w.cited_by_count || 0,
    a: displayAuthors(w.authors) || (w.title ? w.title.slice(0, 24) : w.id),
    v: w.venue || "",
    t: w.title || w.id,
    doi: w.doi,
    arxiv_id: w.arxiv_id,
    doc_id: w.doc_ids.length > 0 ? (w.doc_ids[0] as string) : null,
  };
}

export function buildGraph(store: LibraryStore, outputDir: string): GraphData {
  const oaid2node = new Map<string, string>();
  const nodes: GraphNode[] = [];
  for (const w of store.works) {
    nodes.push(savedNode(w));
    if (w.openalex_id) oaid2node.set(w.openalex_id, w.id);
  }

  const linkSeen = new Set<string>();
  const links: Edge[] = [];
  const addLink = (e: Edge): void => {
    if (!linkSeen.has(edgeKey(e))) {
      linkSeen.add(edgeKey(e));
      links.push(e);
    }
  };

  // saved → saved, from the OpenAlex references already stored on the works
  for (const w of store.works) {
    for (const r of w.referenced_works) {
      const tgt = oaid2node.get(r);
      if (tgt && tgt !== w.id) addLink([w.id, tgt]);
    }
  }
  // saved → saved, from the ingested bibliographies (guaranteed regardless of OpenAlex)
  for (const e of offlineEdges(store, outputDir)) addLink(e);

  links.sort(compareEdges);
  return { version: CURRENT_GRAPH_VERSION, nodes, links: links.map(([a, b]) => [a, b]) };
}

// --------------------------------------------------------------------------- //
// API projections
// --------------------------------------------------------------------------- //

export function workToRef(w: Work): LibraryRef {
  // required fields (the TS LibraryRef declares these non-optional) are always
  // present — never stripped — so the frontend can rely on e.g. r.authors.
  const ref: LibraryRef = {
    id: w.id,
    title: w.title || w.id,
    authors: displayAuthors(w.authors),
    year: w.year || 0,
    venue: w.venue || "",
    type: w.type as "article" | "conf", // verbatim (contract enum; only these two occur)
    cite: (pyOr(w.cite_key) as string | undefined) ?? slug(w.id),
    tags: w.tags,
    pdf: w.doc_ids.length > 0,
    read: w.read,
    star: w.star,
  };
  const acq = w.acquisition ?? {};
  const optional: Record<string, unknown> = {
    abstract: w.abstract,
    note: w.note,
    doi: w.doi,
    arxiv_id: w.arxiv_id,
    // Stage 14: presence enables the web "explore related papers" action
    bibcode: w.bibcode,
    doc_id: w.doc_ids.length > 0 ? w.doc_ids[0] : null,
    // all versions (Stage 7 MS3): doc_ids[0] IS the main doc (same as doc_id)
    doc_ids: w.doc_ids.length > 0 ? w.doc_ids : null,
    citedBy: w.cited_by_count,
    label: w.label,
    journal: w.journal,
    // acquisition: which full-text source is planned + whether it's ready now
    source: acq.chosen,
    sourceLabel: acq.chosen_label,
    sourceStatus: acq.status,
    sourceReady: acq.ready,
    // bot-walled & no auto source → the UI offers a PDF upload
    needs_upload: pyOr(acq.needs_upload) ?? null,
    // resolution: which provider supplied the citation count (ads/crossref/openalex)
    resolvedBy: w.resolution?.count,
  };
  for (const [k, v] of Object.entries(optional)) {
    if (v !== null && v !== undefined && v !== "") {
      (ref as unknown as Record<string, unknown>)[k] = v;
    }
  }
  return ref;
}

// --------------------------------------------------------------------------- //
// Incremental update (Stage 13)
// --------------------------------------------------------------------------- //

/**
 * Merge one manually added work into an existing graph (Stage 13; Stage 14:
 * saved-only): its saved node plus every locally computable saved↔saved edge
 * (inbound from saved works whose `referenced_works` contain it, outbound to
 * saved works it cites). No OpenAlex candidate fetch for suggested
 * neighbours — those no longer exist.
 *
 * This is a best-effort immediate view, not a rebuild: offline
 * (parsed-bibliography) edges are only computed by {@link buildGraph}, so the
 * next `library build`/refresh converges the graph to authoritative. Callers
 * treat failure as non-fatal (the work is already saved; the graph catches up
 * on the next refresh).
 */
export function mergeWorkIntoGraph(graph: GraphData, store: LibraryStore, w: Work): GraphData {
  const nodes: GraphNode[] = graph.nodes.filter((n) => n.id !== w.id);
  nodes.push(savedNode(w));

  const oaid2node = new Map<string, string>();
  for (const x of store.works) if (x.openalex_id) oaid2node.set(x.openalex_id, x.id);

  const links: [string, string][] = graph.links.map(([a, b]) => [a, b]);
  const seen = new Set(links.map(edgeKey));
  const add = (e: [string, string]): void => {
    if (e[0] !== e[1] && !seen.has(edgeKey(e))) {
      seen.add(edgeKey(e));
      links.push(e);
    }
  };

  // inbound: saved works that cite the new one
  if (w.openalex_id) {
    for (const x of store.works) {
      if (x.id !== w.id && x.referenced_works.includes(w.openalex_id)) add([x.id, w.id]);
    }
  }
  // outbound: saved works the new one cites
  for (const r of w.referenced_works) {
    const tgt = oaid2node.get(r);
    if (tgt) add([w.id, tgt]);
  }

  links.sort(compareEdges);
  return { version: CURRENT_GRAPH_VERSION, nodes, links };
}
